src/Layer/C3DTilesLayer.js

```diff
@@ -3,6 +3,270 @@ import GeometryLayer from 'Layer/GeometryLayer';
+     */
+    setUserData(key, value, updateStyle = true) {
+        this.#userData[key] = value;
+
+        if (updateStyle) {
+            this.#layer._styleController.updateStyle(this.#layer, [this._tileId]);
+        }
+    }
+
+    /**
+     *
+     * @param {string} key - key of the userData object
+     * @returns {object}
+     */
+    getUserData(key) {
+        return this.#userData[key];
+    }
+
+    /**
+     * get info (no setter since info should not change at runtime)
+     * @returns {object} - batchTable info
+     */
+    getInfo() {
+        return this.#info;
+    }
+
+    /**
+     * Use this method to modify user data of an array of batchelements (updateStyle only one time)
+     *
+     * @param {Array<C3DTileFeature>} c3DTileFeatures - c3DTileFeature to update userdata
+     * @param {string} key - the key of user data to set
+     * @param {object} value - the value of key to set
+     */
+    static setUserData(c3DTileFeatures, key, value) {
+        if (!c3DTileFeatures.length) {
+            return;
+        }
+        // record what layer + tile ids have to be updated
+        const updateStyleInfos = [];
+        c3DTileFeatures.forEach((b) => {
+            const resultFilter = updateStyleInfos.filter(el => el.layer === b.#layer);
+            if (resultFilter.length) {
+                // this layer has already been added
+                const updateStyleInfo = resultFilter[0];
+                if (!updateStyleInfo.tileIds.includes(b._tileId)) {
+                    // tile id of this layer has not been added
+                    updateStyleInfo.tileIds.push(b._tileId);
+                }
+                // already register for an update
+            } else {
+                // register this layer + its tile id
+                updateStyleInfos.push({
+                    layer: b.#layer,
+                    tileIds: [b._tileId],
+                });
+            }
+
+            b.setUserData(key, value, false);
+        });
+
+        // update style
+        updateStyleInfos.forEach((usi) => {
+            usi.layer._styleController.updateStyle(usi.layer, usi.tileIds);
+        });
+    }
+}
+
+class StyleController {
+    constructor() {
+        /** @type {Map<string,THREE.MeshStandardMaterial} */
+        this._fillColorMaterialsBuffer = new Map();
+    }
+
+    /**
+     *
+     * @param {C3DTilesLayer} c3DTilesLayer - C3DTilesLayer to update style
+     * @param {Array<number>|null} [allowTileIdList] - tile ids to allow in updateStyle computation if null all tile id are computed
+     * @returns {boolean} true if style updated false otherwise
+     */
+    updateStyle(c3DTilesLayer, allowTileIdList = null) {
+        if (!c3DTilesLayer._style) {
+            return false;
+        }
+
+        const currentMaterials = [];// list materials used for this update
+        for (const [tileId, c3DTileFeatures] of c3DTilesLayer.tilesC3DTileFeatures) {
```

Comment:
not sure to understand what you mean

Reply:
For each `tileId` you get the `Object3D` of `layer.object3d` with the `tileId` property, meaning that for each step of the for loop (for each `tileId`) you traverse the tree once (maybe not fully) to get the right `Object3D`. Another way would be to traverse `layer.object3d` once and apply the style modifications only to the right tiles. So instead of:

```
for (const [tileId, c3DTileFeatures] of this.tilesC3DTileFeatures) {
            const tileContent = this.object3d.getObjectByProperty('tileId', tileId);
            // Do stuff on tileContent
}
```

something like:

```
this.object3d.traverse((obj) => {
     if (obj.tileId && this.tilesC3DTileFeatures.has(obj.tileId)) { 
          // Do stuff on tileContent
     }
});
```

I think it should be more efficient in many cases but I may be wrong, WDYT?

Reply:
Indeed you right, I did the modification